c10/core/impl/COW.cpp

```diff
@@ -129,9 +130,34 @@ C10_API void materialize_cow_storage(StorageImpl& storage) {
     // the context ensured they finished before giving us the result.
     std::unique_ptr<void, DeleterFnPtr> data =
         std::get<cow::COWDeleterContext::LastReference>(std::move(result));
-    TORCH_INTERNAL_ASSERT(data.get() == data_ptr.get());
-    new_data_ptr = DataPtr(
+    // data is a UnifiedMemoryDataPtrContext.
+    if (data.get_deleter() == c10::impl::cow::unified_memory_data_ptr_ctx_deleter) {
+      auto* unified_data_ptr_ctx =
+        reinterpret_cast<c10::impl::cow::UnifiedMemoryDataPtrContext*>(data.get());
+
+      // Same device, owning the pointer.
+      if (data_ptr.get() == unified_data_ptr_ctx->get_original_data_ctx()) {
+        auto original_data = unified_data_ptr_ctx->move_original_data_ctx();
+        new_data_ptr = DataPtr(
+          data_ptr.get(), original_data.release(), original_data.get_deleter(), data_ptr.device());
+      } else {
+        // Different devices copy
+        // TODO(Frank): Technically, it's possible to have CPUTensor owns
+        // MPS buffer, but it will be troublesome if we want to do other things.
+        // Eg. COW-again with the CPUTensor owning the MPS buffer, which means
+        // we will have logic to determine whether the data owned by the Tensor
+        // is created by the device it's on. It's possible, but will be
+        // much complicated, since the current assumption is, if the device
+        // is not COW'ed, data_ptr.get() == data.get().
+        // For now we will perform a copy.
+        new_data_ptr = storage.allocator()->clone(data_ptr.get(), storage.nbytes());
```

Comment:
So in the case where `mps_tensor = cpu_tensor.to("mps")` is followed by a write to `cpu_tensor` and then a write to `mps_tensor`, both writes will initiate a copy, right? In other words, we're replacing the single immediate copy with two lazy copies. Do the benefits outweigh the costs?

Reply:
This is correct case where the benefits will outweigh the cost. Do you have any suggestions on how this problem could be resolved other than let the `mps_tensor` owning the cpu memory? Or do you think we should figure out a way to support tensor owning a chunk of memory that's not created by the device allocator?

Reply:
First, maybe it's ultimately ok to have the extra copy--I don't know. It depends on what the typical use case is expected to be.

> Do you have any suggestions on how this problem could be resolved other than let the mps_tensor owning the cpu memory?

I have a thought, though I'm not sure if it will be a good idea in the end. In the specific case I gave above, during the first write, to `cpu_tensor`, we could copy the CPU data to MPS and make `mps_tensor` point to it. Then `cpu_tensor` would take ownership of the CPU data when it materializes. Then, when `mps_tensor` is written, the data would already be on MPS, so it can materialize without copy.

A potential way to implement that would be to add a layer of indirection between `UniqueVoidPtr` and the underlying data, let's call it `Buffer`. After `mps_tensor = cpu.tensor("mps")` is called, `cpu_tensor` and `mps_tensor` would initially point to the same `Buffer`, and the buffer would point to the original CPU data. When `cpu_tensor` is written, it would copy the CPU data to MPS and change `Buffer` to point to the new MPS data.

But this is just a brainstorm, maybe there is a better way. This idea would also have to handle more complex cases, with more than just two tensors.

> Or do you think we should figure out a way to support tensor owning a chunk of memory that's not created by the device allocator?

It seems like that would be kind of against the whole purpose of the the `mps_tensor = cpu_tensor.to("mps")` call, wouldn't it? To me it seems like it would be odd for `mps_tensor` to still point to CPU data even after materialization. But I don't know, is it really possible/desirable to do that?